File: [K]graph/[K]graph-bfs/733-flood-fill.py
from collections import deque
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        rows, cols = len(image), len(image[0])
        old_color = image[sr][sc]
        visited = set()
        queue = deque([(sr, sc)])
        visited.add((sr, sc))
        while queue:
            r, c = queue.popleft()
            image[r][c] = color
            for next_r, next_c in [(r+1, c), (r-1, c), (r, c+1), (r, c-1)]:
                if next_r not in range(rows) or next_c not in range(cols) or (next_r, next_c) in visited or image[next_r][next_c] != old_color:
                    continue
                queue.append((next_r, next_c)) 
                visited.add((next_r, next_c))
        return image

# time O(RC)
# space O(RC)
# using graph and bfs with single source

File: [K]graph/[K]graph-bfs/733-flood-fill.java
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        int oldColor = image[sr][sc];
        if (oldColor == color) {
            return image;
        }
        int rows = image.length;
        int cols = image[0].length;
        ArrayDeque<int[]> queue = new ArrayDeque<>();
        HashSet<Integer> visited = new HashSet<>();
        queue.offer(new int[]{sr, sc});
        visited.add(sr * cols + sc);
        while (!queue.isEmpty()) {
            int[] rc = queue.poll();
            int r = rc[0];
            int c = rc[1];
            image[r][c] = color;
            for (int[] d: new int[][]{{1, 0}, {- 1, 0}, {0, 1}, {0, - 1}}) {
                int nextR = r + d[0];
                int nextC = c + d[1];
                if (0 <= nextR && nextR < rows && 0 <= nextC && nextC < cols && !visited.contains(nextR * cols + nextC) && image[nextR][nextC] == oldColor) {
                    queue.offer(new int[]{nextR, nextC});
                    visited.add(nextR * cols + nextC);
                }
            }
        }
        return image;
    }
}

// time O(RC)
// space O(RC)
// using graph and bfs with single source